lib/ash_authentication/strategies/api_key/actions.ex

```diff
@@ -0,0 +1,75 @@
+defmodule AshAuthentication.Strategy.ApiKey.Actions do
+  @moduledoc """
+  Actions for the magic link strategy.
+
+  Provides the code interface for working with user resources for providing
+  magic links.
+  """
+
+  alias Ash.{Query, Resource}
+  alias AshAuthentication.{Errors, Info, Strategy.ApiKey}
+
+  @doc """
+  Attempt to sign a user in via magic link.
+  """
+  @spec sign_in(ApiKey.t(), map, keyword) ::
+          {:ok, Resource.record()} | {:error, Errors.AuthenticationFailed.t()}
+  def sign_in(strategy, params, options) do
+    options =
+      options
+      |> Keyword.put_new_lazy(:domain, fn -> Info.domain!(strategy.resource) end)
+
+    strategy.resource
+    |> Query.new()
+    |> Query.set_context(%{private: %{ash_authentication?: true}})
+    |> Query.for_read(strategy.sign_in_action_name, params, options)
+    |> Ash.read()
+    |> case do
+      {:ok, [user]} ->
+        {:ok, user}
+
+      {:ok, []} ->
+        {:error,
+         Errors.AuthenticationFailed.exception(
+           strategy: strategy,
+           caused_by: %{
+             module: __MODULE__,
+             strategy: strategy,
+             action: :sign_in,
+             message: "Query returned no users"
+           }
+         )}
+
+      {:ok, _users} ->
+        {:error,
+         Errors.AuthenticationFailed.exception(
+           strategy: strategy,
+           caused_by: %{
+             module: __MODULE__,
+             strategy: strategy,
+             action: :sign_in,
+             message: "Query returned too many users"
```

Comment:
I would not differentiate between no users or too many users.

If two users use the same insecure token, we don't want them to know that deleting theirs will let them access the other one.

Also there should probably be a unique index by default preventing this in the first place.

Reply:
This error isn't returned to the end user, it is only used internally (and yeah there is an identity added to protect about this).

But I could also hide this information even from the developer if that is more secure?

Reply:
If it's only internal, it's probably fine.